cistem/protocols/protocol_ts_resample.py

```diff
@@ -144,86 +124,35 @@ def runTsResample(self, tsObjId):
+        outTsSet.copyInfo(inputTs)
+        outTsSet.setSamplingRate(self._getOutputSampling())
+        outTsSet.copyItems(inputTs, updateTiCallback=self._updateTi)
 
-    def createOutputStep(self, tsObjId):
-        ts = self.inputSetOfTiltSeries.get()[tsObjId]
+        self._defineOutputs(**{OUTPUT_TS_NAME: outTsSet})
+        self._defineTransformRelation(self.inputSetOfTiltSeries, outTsSet)
 
-        firstItem = ts.getFirstItem()
-        tsFile = firstItem.getFileName()
-        tsBaseName = removeBaseExt(tsFile)
-        tsExt = getExt(tsFile)
-
-        tsOutName = self.getWorkingDir() + '/' + OUTPUT_DIR + '/' + \
-            tsBaseName + '_resampled' + tsExt
-
-        tsId = ts.getTsId()
-
-        oldSamplingRate = ts.getSamplingRate()
-        oldXsize = ts.getDim()[0]
-        self.samplingRate = oldSamplingRate * oldXsize / float(self.newXsize)
-
-        outputSetOfTiltSeries = self.getOutputSetOfTiltSeries()
-
-        newTs = TiltSeries(tsId=tsId)
-        newTs.copyInfo(ts)
-        outputSetOfTiltSeries.append(newTs)
-
-        newTs.setSamplingRate(self.samplingRate)
-
-        for index, ti in enumerate(ts):
-            newTi = tomoObj.TiltImage()
-            newTi.copyInfo(ti, copyId=True)
-            newTi.setLocation(index + 1, tsOutName)
-
-            if ti.hasTransform():
-                newTi.setTransform(ti.getTransform())
-
-            newTi.setSamplingRate(self.samplingRate)
-
-            newTs.append(newTi)
-
-        ih = ImageHandler()
-        x, y, z, _ = ih.getDimensions(newTs.getFirstItem().getFileName())
-        newTs.setDim((x, y, z))
-        newTs.write(properties=False)
-
-        outputSetOfTiltSeries.update(newTs)
-        outputSetOfTiltSeries.updateDim()
-        outputSetOfTiltSeries.write()
-
-        self._store()
-
-    def closeStreamStep(self):
-        self.getOutputSetOfTiltSeries().setStreamState(Set.STREAM_CLOSED)
-
-        self._store()
+    # --------------------------- INFO functions -----------------------------------
+    def _citations(self):
+        return ['Grant2018']
 
     # --------------------------- UTILS functions -------------------------------
-    def getOutputSetOfTiltSeries(self):
-        if hasattr(self, "outputSetOfTiltSeries"):
-            self.outputSetOfTiltSeries.enableAppend()
-        else:
-            outputSetOfTiltSeries = self._createSetOfTiltSeries()
-            outputSetOfTiltSeries.copyInfo(self.inputSetOfTiltSeries.get())
-            outputSetOfTiltSeries.setDim(self.inputSetOfTiltSeries.get().getDim())
-            outputSetOfTiltSeries.setSamplingRate(self.samplingRate)
-            outputSetOfTiltSeries.setStreamState(Set.STREAM_OPEN)
-            self._defineOutputs(outputSetOfTiltSeries=outputSetOfTiltSeries)
-            self._defineSourceRelation(self.inputSetOfTiltSeries, outputSetOfTiltSeries)
-        return self.outputSetOfTiltSeries
+    def _updateTi(self, j, ts, ti, tsOut, tiOut):
```

Comment:
This is the key step that was confusing me especially in the tilt-series case. Now it's much cleaner!

Reply:
We cannot remove the unused arguments (namely j, ts, tsOut) because the function call needs to match this below?
https://github.com/scipion-em/scipion-em-tomo/blob/01bb324a2f6bc3dda4c57aee6853570af65719b0/tomo/objects.py#L794

that's very cryptic... I would never have known that by myself.